streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -230,8 +234,19 @@ private void emitNonJoinedOuterRecords(
                     sharedTimeTracker.minTime = timestamp;
 
                     // Skip next records if window has not closed
-                    if (timestamp + joinAfterMs + joinGraceMs >= sharedTimeTracker.streamTime) {
-                        break;
+                    final long outerJoinLookBackTimeMs = getOuterJoinLookBackTimeMs(timestampedKeyAndJoinSide);
+                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs >= sharedTimeTracker.streamTime) {
+                        if (timestampedKeyAndJoinSide.isLeftSide()) {
+                            outerJoinLeftBreak = true; // there are no more candidates to emit on left-outerJoin-side
+                        } else {
+                            outerJoinRightBreak = true; // there are no more candidates to emit on right-outerJoin-side
+                        }
+                        if (outerJoinLeftBreak && outerJoinRightBreak) {
```

Comment:
Hi,

Seems like `outerJoinLeftBreak && outerJoinRightBreak` is always false.
Doesn't this break the behaviro described in the comment on top of this block? `// Skip next records if window has not closed`?

Reply:
Probably these two lines need to be outside the while loop:
```
                    boolean outerJoinLeftBreak = false;
                    boolean outerJoinRightBreak = false;
```

Reply:
See: https://github.com/apache/kafka/pull/15510